airbyte-commons-worker/src/main/java/io/airbyte/workers/internal/sync_persistence/SyncPersistenceImpl.java

```diff
@@ -0,0 +1,176 @@
+import java.util.UUID;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import lombok.extern.slf4j.Slf4j;
+
+@Slf4j
+public class SyncPersistenceImpl implements SyncPersistence {
+
+  final long RUN_IMMEDIATELY = 0;
+  final long FLUSH_TERMINATION_TIMEOUT_IN_SECONDS = 60;
+
+  private final UUID connectionId;
+  private final StateApi stateApi;
+  private final StateAggregatorFactory stateAggregatorFactory;
+
+  private StateAggregator stateBuffer;
+  private StateAggregator stateToFlush;
+  private final ScheduledExecutorService stateFlushExecutorService;
+  private ScheduledFuture<?> stateFlushFuture;
+  private final long stateFlushPeriodInSeconds;
+
+  public SyncPersistenceImpl(final UUID connectionId,
+                             final StateApi stateApi,
+                             final StateAggregatorFactory stateAggregatorFactory,
+                             final ScheduledExecutorService scheduledExecutorService,
+                             long stateFlushPeriodInSeconds) {
+    this.connectionId = connectionId;
+    this.stateApi = stateApi;
+    this.stateAggregatorFactory = stateAggregatorFactory;
+    this.stateFlushExecutorService = scheduledExecutorService;
+    this.stateBuffer = this.stateAggregatorFactory.create();
+    this.stateFlushPeriodInSeconds = stateFlushPeriodInSeconds;
+  }
+
+  @Override
+  public void persist(AirbyteStateMessage stateMessage) {
+    stateBuffer.ingest(stateMessage);
+    startBackgroundFlushStateTask();
+  }
+
+  private void startBackgroundFlushStateTask() {
+    if (stateFlushFuture != null) {
+      return;
+    }
+
+    // Making sure we only start one of background flush task
+    synchronized (this) {
+      if (stateFlushFuture == null) {
+        stateFlushFuture =
+            stateFlushExecutorService.scheduleAtFixedRate(this::flushState, RUN_IMMEDIATELY, stateFlushPeriodInSeconds, TimeUnit.SECONDS);
+      }
+    }
+  }
+
+  /**
+   * Stop background data flush thread and attempt to flush pending data
+   *
+   * If there is already flush in progress, wait for it to terminate. If it didn't terminate during
+   * the allocated time, we exit rather than attempting a concurrent write that could lead to
+   * non-deterministic behavior.
+   *
+   * For the final flush, we will retry in case of failures since there is no more "scheduled" attempt
+   * after this.
+   */
+  @Override
+  public void close() {
+    // stop the buffered refresh
+    stateFlushExecutorService.shutdown();
+
+    // Wait for previous running task to terminate
+    try {
+      final boolean terminated = stateFlushExecutorService.awaitTermination(FLUSH_TERMINATION_TIMEOUT_IN_SECONDS, TimeUnit.SECONDS);
+      if (!terminated) {
+        // Ongoing flush failed to terminate within the allocated time
+        log.info("Pending persist operation took too long to complete, most recent states may have been lost");
```

Comment:
Nit: this should be an error log

Reply:
I thought about it, I think it should be monitored, my argument for not logging an error is that this isn't actionable and it draws attention towards the wrong things. The outcome of a failure here is that we are going to have duplicated data in the raw tables which is going to happen regardless without a bigger change in how we approach state messages.